librosa/core/spectrum.py

```diff
@@ -759,6 +761,197 @@ def perceptual_weighting(S, frequencies, **kwargs):
+    >>> odf_ac = librosa.autocorrelate(odf, max_size=10 * sr // 512)
+    >>> odf_ac_scale = librosa.fmt(librosa.util.normalize(odf_ac), n_fmt=512)
+    >>> # Plot the results
+    >>> plt.figure()
+    >>> plt.subplot(3, 1, 1)
+    >>> plt.plot(odf, label='Onset strength')
+    >>> plt.axis('tight')
+    >>> plt.xlabel('Time (frames)')
+    >>> plt.xticks([])
+    >>> plt.legend(frameon=True)
+    >>> plt.subplot(3, 1, 2)
+    >>> plt.plot(odf_ac, label='Onset autocorrelation')
+    >>> plt.axis('tight')
+    >>> plt.xlabel('Lag (frames)')
+    >>> plt.xticks([])
+    >>> plt.legend(frameon=True)
+    >>> plt.subplot(3, 1, 3)
+    >>> plt.semilogy(np.abs(odf_ac_scale), label='Scale transform magnitude')
+    >>> plt.axis('tight')
+    >>> plt.xlabel('scale coefficients')
+    >>> plt.legend(frameon=True)
+    >>> plt.tight_layout()
+    """
+
+    n = y.shape[axis]
+
+    if n < 3:
+        raise ParameterError('y.shape[{:}]=={:} < 3'.format(axis, n))
+
+    if t_min <= 0:
+        raise ParameterError('t_min must be a positive number')
+
+
+    if n_fmt is None:
+        if over_sample < 1:
+            raise ParameterError('over_sample must be >= 1')
+
+        # The base is the maximum ratio between adjacent samples
+        # Since the sample spacing is increasing, this is simply the
+        # ratio between the positions of the last two samples: (n-1)/(n-2)
+        log_base = np.log(n - 1) - np.log(n - 2)
+
+        n_fmt = int(np.ceil(over_sample * (np.log(n - 1) - np.log(t_min)) / log_base))
+
+    elif n_fmt < 3:
+        raise ParameterError('n_fmt=={:} < 3'.format(n_fmt))
+    else:
+        log_base = (np.log(n_fmt - 1) - np.log(n_fmt - 2)) / over_sample
+
+    if not np.all(np.isfinite(y)):
+        raise ParameterError('y must be finite everywhere')
+
+    base = np.exp(log_base)
+    # original grid: signal covers [0, 1).  This range is arbitrary, but convenient.
+    # The final sample is positioned at (n-1)/n, so we omit the endpoint
+    x = np.linspace(0, 1, num=n, endpoint=False)
+
+    # build the interpolator
+    f_interp = scipy.interpolate.interp1d(x, y, kind=kind, axis=axis)
+
+    # build the new sampling grid
+    # exponentially spaced between t_min/n and 1 (exclusive)
+    # we'll go one past where we need, and drop the last sample
+    # When over-sampling, the last input sample contributions n_over samples.
+    # To keep the spacing consistent, we over-sample by n_over, and then
+    # trim the final samples.
+    n_over = int(np.ceil(over_sample))
+    x_exp = np.logspace((np.log(t_min) - np.log(n)) / log_base,
+                        0,
+                        num=n_fmt + n_over,
+                        endpoint=False,
+                        base=base)[:-n_over]
+
+    # Clean up any rounding errors at the boundaries of the interpolation
+    # The interpolator gets angry if we try to extrapolate, so clipping is necessary here.
+    x_exp = np.clip(x_exp, float(t_min) / n, x[-1])
```

Comment:
I'm glad that you picked up on this point, since it's certainly the most head-scratching part of this implementation.

Dan and I went back and forth on this in the PR thread earlier.  I think we all agree that clipping is generally terrible.  The reason that it's included here is to resolve floating point precision errors at the interpolation boundaries.  It's not that we're extrapolating exactly, it's that the sample point generator is fuzzing slightly out of the original sample domain --- ``[t_min, (n-1)/n]`` --- and we need to lock it back down.

Using a mask (throwing out these points, if I understand correctly) seems incorrect here, since it would throw off the sample count.  Or maybe there's a more clever way to do this?